### PR TITLE
p2p: improve PeerManager prototype

### PR DESCRIPTION
**Pull request by erikgrinaker**
This improves the prototype peer manager by:

* Exporting `PeerManager`, making it accessible by e.g. reactors.
* Replacing `Router.SubscribePeerUpdates()` with `PeerManager.Subscribe()`.
* Tracking address/peer connection statistics, and retrying dial failures with exponential backoff.
* Prioritizing peers, with persistent peers configuration.
* Limiting simultaneous connections.
* Evicting peers and upgrading to higher-priority peers.
* Tracking peer heights, as a workaround for legacy shared peer state APIs.

This is getting to a point where we need to determine precise semantics and implement tests, so we should figure out whether it's a reasonable abstraction that we want to use. The main questions are around the API model (i.e. synchronous method calls with the router polling the manager, vs. an event-driven model using channels, vs. the peer manager calling methods on the router to connect/disconnect peers), and who should have the responsibility of managing actual connections (currently the router, while the manager only tracks peer state).

### CODE REVIEW

File: p2p/peer.go [hunk 1]
```diff
@@ -490,35 +748,76 @@ func (s *peerStore) List() ([]*peerInfo, error) {
...
+func (s *peerStore) Ranked() ([]*peerInfo, error) {
+	peers, err := s.List()
+	if err != nil {
+		return nil, err
+	}
+	sort.Slice(peers, func(i, j int) bool {
+		// FIXME: If necessary, consider precomputing scores before sorting,
+		// to reduce the number of Score() calls.
+		return peers[i].Score() < peers[j].Score()
+	})
```
Comment (cmwaters) [inline p2p/peer.go:770]: It seems to me from looking at the example here: https://golang.org/pkg/sort/#Slice, that this will sort in ascending order but what we want is descending order right i.e. `(better peers first)`

Comment (erikgrinaker) [reply p2p/peer.go:770]: Ah, of course, thanks!

File: p2p/peer.go [hunk 2]
```diff
@@ -235,25 +239,91 @@ type PeerUpdate struct {
...
+	// know or learn about better-scored peers (e.g. configured persistent
+	// peers) that we are not connected too, then we can probe these peers by
+	// using up to MaxConnectedUpgrade connections, and once connected evict the
+	// lowest-scored connected peers. This also works for inbound connections,
+	// i.e. if a higher-scored peer attempts to connect to us, we can accept
+	// the connection and evict a lower-scored peer.
+	MaxConnectedUpgrade uint16
+
+	// MinRetryTime is the minimum time to wait between retries. Retry times
+	// double for each retry, up to MaxRetryTime. 0 disables retries.
```
Comment (melekes) [inline p2p/peer.go:291]: why not exponential retry?

File: p2p/peer.go [hunk 3]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
...
+// retry settings in PeerManagerOptions. If MinRetryTime is 0, this returns
+// MaxInt64 (i.e. an infinite retry delay, effectively disabling retries).
+func (m *PeerManager) retryDelay(peer *peerInfo, failures uint32) time.Duration {
+	if failures == 0 {
+		return 0
+	}
+	if m.options.MinRetryTime == 0 {
+		return time.Duration(math.MaxInt64)
+	}
+	delay := m.options.MinRetryTime * time.Duration(math.Pow(2, float64(failures)))
```
Comment (melekes) [inline p2p/peer.go:470]: will MaxInt64 overflow here?

File: p2p/peer.go [hunk 4]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
...
+		return fmt.Errorf("peer %q was removed while dialing", peerID)
 	}
+	m.connected[peerID] = true
 
-	if m.connected[peerID] {
-		return fmt.Errorf("peer %v is already connected", peerID)
+	now := time.Now().UTC()
+	peer.LastConnected = now
+	if addressInfo := peer.LookupAddressInfo(address); addressInfo != nil {
+		addressInfo.DialFailures = 0
```
Comment (melekes) [inline p2p/peer.go:533]: should we reset `LastDialFailure` too?

File: p2p/peer.go (hunk 2 above)
Comment (erikgrinaker) [reply p2p/peer.go:291]: This is exponential retry, using base-2 exponentials (which is a typical base to use for exponential backoff). If we use `MinRetryTime` as the base it blows up real fast, with e.g. "100 milliseconds" they develop like this:

* Base 2: 100, 200, 400, 800, 1600, 3200, etc.
* Base MinRetryTime: 100, 10000, 10000000, etc.

File: p2p/peer.go (hunk 4 above)
Comment (erikgrinaker) [reply p2p/peer.go:533]: I think it might be useful to keep it around, e.g. for more advanced peer quality scoring.

File: p2p/peer.go (hunk 3 above)
Comment (erikgrinaker) [reply p2p/peer.go:470]: Yeah, we need to be careful with stuff like this. It's still a prototype, so not dealing with it now, but will handle this one way or another when writing tests.

File: p2p/peer.go [hunk 5]
```diff
@@ -235,25 +239,91 @@ type PeerUpdate struct {
 // - Disconnected: peer disconnects, unmarking as connected and broadcasts a
 //   PeerStatusDown peer update.
 //
+// If we need to evict a peer, typically because we have connected to additional
+// higher-scored peers and need to shed lower-scored ones,, the flow is as follows:
```
Comment (alexanderbez) [inline p2p/peer.go]:
```suggestion
// higher-scored peers and need to shed lower-scored ones, the flow is as follows:
```

File: p2p/peer.go [hunk 6]
```diff
@@ -235,25 +239,91 @@ type PeerUpdate struct {
...
-func newPeerManager(store *peerStore) *peerManager {
-	return &peerManager{
-		store:         store,
+// PeerManagerOptions specifies options for a PeerManager.
+type PeerManagerOptions struct {
+	// PersistentPeers are peers that we want to maintain persistent connections
+	// to. These will be scored higher than other peers, and if
+	// MaxConnectedUpgrade is non-zero any lower-scored peers will be evicted if
+	// necessary to make room for these.
+	PersistentPeers []NodeID
```
Comment (alexanderbez) [inline p2p/peer.go:272]: What happens when `len(PersistentPeers) > MaxConnected`?

File: p2p/peer.go [hunk 7]
```diff
@@ -235,25 +239,91 @@ type PeerUpdate struct {
...
+	// MaxRetryTime is the maximum time to wait between retries. 0 means
+	// no maximum, in which case the retry time will keep doubling.
+	MaxRetryTime time.Duration
+
+	// MaxRetryTimePersistent is the maximum time to wait between retries for
+	// peers listed in PersistentPeers. 0 uses MaxRetryTime instead.
+	MaxRetryTimePersistent time.Duration
+
+	// FuzzRetryTime is the upper bound of a random interval added to
+	// retry times, to avoid thundering herds. 0 disables fuzzing.
```
Comment (alexanderbez) [inline p2p/peer.go:303]: Maybe use the "jitter" nomenclature here instead? Or maybe just reference https://en.wikipedia.org/wiki/Jitter or something like:

```
Clients will retry failed calls, by waiting a specific amount of time between consecutive retries. In order to avoid the thundering herd problem, jitter can be purposefully introduced, in order to break the synchronization across the clients thereby avoiding collisions.
```

File: p2p/peer.go [hunk 8]
```diff
@@ -235,25 +239,91 @@ type PeerUpdate struct {
...
+	// to. These will be scored higher than other peers, and if
+	// MaxConnectedUpgrade is non-zero any lower-scored peers will be evicted if
+	// necessary to make room for these.
+	PersistentPeers []NodeID
+
+	// MaxConnected is the maximum number of connected peers (inbound and
+	// outbound). 0 means no limit.
+	MaxConnected uint16
+
+	// MaxConnectedUpgrade is the maximum number of additional peer connections to
```
Comment (alexanderbez) [inline p2p/peer.go:278]: Sorry if this is obvious, but is this `MaxConnectedUpgrade` new wrt to the existing p2p options? I'm trying to understand why `MaxConnectedUpgrade` is necessary? If we have exhausted `MaxConnected`, why not just evict a peer with the lowest score from that set? Why have an additional upgrade/buffer?

File: p2p/peer.go [hunk 9]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
 
 // DialNext finds an appropriate peer address to dial, and marks it as dialing.
 // The peer will not be returned again until Dialed() or DialFailed() is called
-// for the peer and it is no longer connected.
+// for the peer and it is no longer connected. Returns an empty ID if no
+// appropriate peers are available.
```
Comment (alexanderbez) [inline p2p/peer.go:401]: ...or if we can't make any more connections based on configuration.

File: p2p/peer.go [hunk 10]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
...
 
+// retryDelay calculates a dial retry delay using exponential backoff, based on
+// retry settings in PeerManagerOptions. If MinRetryTime is 0, this returns
+// MaxInt64 (i.e. an infinite retry delay, effectively disabling retries).
+func (m *PeerManager) retryDelay(peer *peerInfo, failures uint32) time.Duration {
+	if failures == 0 {
+		return 0
+	}
+	if m.options.MinRetryTime == 0 {
+		return time.Duration(math.MaxInt64)
```
Comment (alexanderbez) [inline p2p/peer.go:468]: Maybe throw this into a `const`?

File: p2p/peer.go [hunk 11]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
...
+	if peer.Persistent && m.options.MaxRetryTimePersistent > 0 {
+		// We have to check this separately, since we don't want to fall back to
+		// MaxRetryTime in cases where it is smaller than MaxRetryTimePersistent.
+		if delay > m.options.MaxRetryTimePersistent {
+			delay = m.options.MaxRetryTimePersistent
+		}
+	} else if m.options.MaxRetryTime > 0 && delay > m.options.MaxRetryTime {
+		delay = m.options.MaxRetryTime
+	}
+	delay += time.Duration(rand.Int63n(int64(m.options.FuzzRetryTime))) // nolint:gosec
```
Comment (alexanderbez) [inline p2p/peer.go:480]: Should we not create a PRNG object in the manager with a seed (i.e. time.now())?

File: p2p/peer.go [hunk 12]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
...
+			// peers, since they will all have the same or lower score than this
+			// peer (since they're ordered by score via peerStore.Ranked).
+			//
+			// FIXME: There is a race condition here where, if there exists a
+			// single lower-scored peer, we may end up dialing multiple
+			// higher-scored new peers that all expect the same lower-scored
+			// peer to be evicted, causing us to take on too many peers. We may
+			// need to reserve the eviction for this specific peer such that
+			// others can't claim it.
+			if m.options.MaxConnected > 0 &&
```
Comment (alexanderbez) [inline p2p/peer.go:447]: Do we not need to check against `MaxConnectedUpgrade`?

File: p2p/peer.go [hunk 13]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
...
 	m.mtx.Lock()
 	defer m.mtx.Unlock()
 
+	delete(m.dialing, peerID)
+
+	if m.connected[peerID] {
+		return fmt.Errorf("peer %v is already connected", peerID)
+	}
+	if m.options.MaxConnected > 0 &&
+		len(m.connected) >= int(m.options.MaxConnected)+int(m.options.MaxConnectedUpgrade) {
```
Comment (alexanderbez) [inline p2p/peer.go:518]: We have a few places where we check this exact conditional. Perhaps move it to an auxiliary function for better readability?

File: p2p/peer.go [hunk 14]
```diff
@@ -327,102 +397,213 @@ func (m *peerManager) broadcast(peerUpdate PeerUpdate) {
...
 	}
-	if m.connected[peerID] {
-		return fmt.Errorf("peer %q is already connected", peerID)
+
+	// If we're already full (i.e. at MaxConnected), but we allow upgrades (and we
+	// know from the check above that we have upgrade capacity), then we can look
+	// for a lower-scored evictable peer, and if found we can accept this connection
+	// anyway and let EvictNext() evict the lower-scored peer for us.
+	//
+	// FIXME: There is a race condition here where, if there exists a single
```
Comment (alexanderbez) [inline p2p/peer.go:579]: How do you propose we go about fixing this? If you haven't already, let's be sure to mark this in #5670.

File: p2p/peer.go [hunk 15]
```diff
@@ -490,35 +761,76 @@ func (s *peerStore) List() ([]*peerInfo, error) {
 	return peers, nil
 }
 
-// peerInfo contains peer information stored in a peerStore.
+// Ranked returns a list of peers ordered by score (better peers first).
+// Peers with equal scores are returned in an arbitrary order.
```
Comment (alexanderbez) [inline p2p/peer.go:765]: We can sort by NodeID in case of ties for deterministic reasoning.

File: p2p/router.go [hunk 16]
```diff
@@ -241,6 +237,9 @@ func (r *Router) acceptPeers(transport Transport) {
 		default:
 		}
 
+		// FIXME: We may need transports to enforce some sort of rate limiting
```
Comment (alexanderbez) [inline p2p/router.go:240]: Let's be sure to mark this in #5670 if you haven't already 👍

**Comment by erikgrinaker** [comment]
Dumped some thoughts on the PeerManager API model here, in case anyone has opinions: https://github.com/tendermint/tendermint/discussions/5938

Tl;dr `DialNext()` and `EvictNext()` should probably return channels.

File: p2p/peer.go [hunk 17]
```diff
@@ -434,23 +615,113 @@ func (m *peerManager) Ready(peerID NodeID) {
...
+	if err != nil || peer == nil {
+		return 0, err
+	}
+	return peer.Height, nil
+}
+
+// SetHeight stores a peer's height, making it available via GetHeight. If the
+// peer is unknown, it is created.
+//
+// FIXME: This is a temporary workaround for the peer state stored via the
```
Comment (alexanderbez) [inline p2p/peer.go:696]: If you haven't already, let's be sure to mark this in #5670.

File: p2p/router.go [hunk 18]
```diff
@@ -480,16 +479,37 @@ func (r *Router) sendPeer(peerID NodeID, conn Connection, queue queue) error {
...
+		select {
+		case <-r.stopCh:
+			return
+		default:
+		}
+
+		peerID, err := r.peerManager.EvictNext()
+		if err != nil {
+			r.logger.Error("failed to find next peer to evict", "err", err)
+			return
```
Comment (alexanderbez) [inline p2p/router.go:494]: I'm a bit confused as to why this function returns here. If we get an `error` it's because `peerStore#List` returned an error, but do we really want return here instead of just `continue` with a sleep maybe?

**Review by alexanderbez — COMMENTED** [review]
Excellent! I left some remarks and questions.

File: p2p/peer.go (hunk 6 above)
Comment (erikgrinaker) [reply p2p/peer.go:272]: Then we connect to the ones we have slots for. There should probably be a check for this when validating options, but I'm not sure yet if it belongs here or in the `P2PConfig`.

File: p2p/peer.go (hunk 7 above)
Comment (erikgrinaker) [reply p2p/peer.go:303]: Good call -- it's easy to mix this up with fuzz testing.

File: p2p/peer.go (hunk 8 above)
Comment (erikgrinaker) [reply p2p/peer.go:278]: Because if we're full, but we know of better peers, we don't want to evict a lower peer until we've actually tried to dial the new peer, handshaked with it, and made sure it's operational. Only then should we evict the lower peer. Otherwise we'll be evicting peers every time we retry a higher-ranked peer (which may still be offline). While we're doing this, we will for a brief moment be connected to both the old and new peer.

Comment (alexanderbez) [reply p2p/peer.go:278]: > Because if we're full, but we know of better peers, we don't want to evict a lower peer until we've actually tried to dial the better peer, handshaked with it, and made sure it's operational. Only then should we evict the lower peer.

Yup, I understand this. I'm just trying to make the connection with that logic and the use of `MaxConnectedUpgrade`. Can't we still evict and perform that logic and just use `MaxConnected`?

Comment (erikgrinaker) [reply p2p/peer.go:278]: `MaxConnectedUpgrade` is basically a cap on the number of probes to better peers. Let's assume we're in a network of 200 nodes (100 high 100 low), with a `MaxConnected` of 50, and we're only connected to low-scored peers. We can't evict before we connect to new nodes. How many better peers should we attempt to connect to simultaneously? All of them?

File: p2p/peer.go (hunk 11 above)
Comment (erikgrinaker) [reply p2p/peer.go:480]: Yes, once this is no longer a prototype. This would be needed for testing as well.

File: p2p/peer.go (hunk 12 above)
Comment (erikgrinaker) [reply p2p/peer.go:447]: Not here. If we're connected to 60 peers, with `MaxConnected` of 50, that means we've upgraded 10 peers to higher-quality ones and need to shed the lower-quality 10 peers.

File: p2p/peer.go (hunk 13 above)
Comment (erikgrinaker) [reply p2p/peer.go:518]: Yeah, maybe once we stabilize all the logic -- for now I'd like to keep it in each place, since the code is still in flux and we may need to vary it in each place.

File: p2p/peer.go (hunk 14 above)
Comment (erikgrinaker) [reply p2p/peer.go:579]: Probably by keeping an `upgrading` map that reserves an eviction slot for a higher peer.

File: p2p/peer.go (hunk 15 above)
Comment (erikgrinaker) [reply p2p/peer.go:765]: That's what I initially did, but e.g. in a 100-node network with `MaxConnected=20` then all nodes in the network would only connect to the same 20 peers, which would be bad.

File: p2p/router.go (hunk 18 above)
Comment (erikgrinaker) [reply p2p/router.go:494]: Yeah, that's sort of a general problem -- if we encounter an unexpected error, should we abort or try to recover? For safety, I think it's better to abort.

File: p2p/peer.go (hunk 8 above)
Comment (alexanderbez) [reply p2p/peer.go:278]: Got it. Thank you.

> MaxConnectedUpgrade is basically a cap on the number of probes to better peers.

This clarified it for me. Maybe add this oneliner to the godoc?

File: p2p/peer.go (hunk 15 above)
Comment (alexanderbez) [reply p2p/peer.go:765]: I see. So we kind of want the pseudo-randomness?

File: p2p/router.go (hunk 18 above)
Comment (alexanderbez) [reply p2p/router.go:494]: Yeah because if that errors and we return, we'll never evict again. But I guess since it error'd, future invocations also won't work?

File: p2p/peer.go (hunk 15 above)
Comment (cmwaters) [reply p2p/peer.go:765]: I was thinking a little about this. Is it an ideal characteristic that nodes avoid staying with the same peers even if they have a high rank and have some small element of exploration?

File: p2p/peer.go (hunk 8 above)
Comment (erikgrinaker) [reply p2p/peer.go:278]: Sure!

File: p2p/router.go (hunk 18 above)
Comment (erikgrinaker) [reply p2p/router.go:494]: Yup. I think it's fair to say that an unexpected error here typically implies a disk IO error, and there's no point in trying to continue, but we should probably propagate this upwards. However, I don't think there's currently any way for `Service` implementations to signal error to the main process.

File: p2p/peer.go (hunk 15 above)
Comment (alexanderbez) [reply p2p/peer.go:765]: Great, we can add this to #5670 and resolve this conversation.

File: p2p/peer.go (hunk 10 above)
Comment (erikgrinaker) [reply p2p/peer.go:468]: Yeah, I'm not entirely convinced we even need this -- if we migrate to a channel-based `DialNext()` API then we'd probably/possibly want to register timers for this instead. I'll leave it for now until the API stabilizes a bit.

File: p2p/peer.go (hunk 12 above)
Comment (erikgrinaker) [reply p2p/peer.go:447]: Sorry, misread this -- though this was in `EvictNext()`. Here, we've already checked that we have upgrade capacity at the start of the method, which is also documented in the comment above. So it's sufficient to only check whether all connection slots are full.

File: p2p/peer.go (hunk 14 above)
Comment (erikgrinaker) [reply p2p/peer.go:581]: Tried implementing this, but the logic became a bit more complicated than I'd like. Will revisit this when I attempt the channel-based API, since that might reorganize this logic a bit anyway (i.e. by pushing the peer to the eviction queue immediately instead of relying on `EvictNext` to determine it later).

**Review by alexanderbez — APPROVED** [review]
